application/F3DStarter.cxx

```diff
@@ -165,6 +166,196 @@ class F3DStarter::F3DInternals
     }
   }
 
+  void addOutputImageMetadata(f3d::image& image)
+  {
+    std::stringstream cameraMetadata;
+    {
+      const auto state = Engine->getWindow().getCamera().getState();
+      const auto vec3toJson = [](const std::array<double, 3>& v)
+      {
+        std::stringstream ss;
+        ss << "[" << v[0] << ", " << v[1] << ", " << v[2] << "]";
+        return ss.str();
+      };
+      cameraMetadata << "{\n";
+      cameraMetadata << "  \"pos\": " << vec3toJson(state.pos) << ",\n";
+      cameraMetadata << "  \"foc\": " << vec3toJson(state.foc) << ",\n";
+      cameraMetadata << "  \"up\": " << vec3toJson(state.up) << ",\n";
+      cameraMetadata << "  \"angle\": " << state.angle << "\n";
+      cameraMetadata << "}\n";
+    }
+
+    image.setMetadata("camera", cameraMetadata.str());
+  }
+
+  /**
+   * Substitute the following variables in a filename template:
+   * - `{app}`: application name (ie. `F3D`)
+   * - `{version}`: application version (eg. `2.4.0`)
+   * - `{version_full}`: full application version (eg. `2.4.0-abcdefgh`)
+   * - `{model}`: current model filename without extension (eg. `foo` for `/home/user/foo.glb`)
+   * - `{model.ext}`: current model filename with extension (eg. `foo.glb` for `/home/user/foo.glb`)
+   * - `{model_ext}`: current model filename extension (eg. `glb` for `/home/user/foo.glb`)
+   * - `{date}`: current date in YYYYMMDD format
+   * - `{date:format}`: current date as per C++'s `std::put_time` format
+   * - `{n}`: auto-incremented number to make filename unique
+   * - `{n:2}`, `{n:3}`, ...: zero-padded auto-incremented number to make filename unique
+   */
+  std::filesystem::path applyFilenameTemplate(
+    const std::string& templateString, size_t maxNumberingAttempts = 1000000)
```

Comment:
not a big fan of having a param that we do not "use" apart the default value. should be a const expr instead.